# Euler Project: Problems from 1 - 50

In [1]:
import numpy as np
from math import ceil
import math
# Allow several prints in one cell
# from IPython.core.interactiveshell import InteractiveShell
# InteractiveShell.ast_node_interactivity = "all"
from math import factorial
import sympy
from sympy import sieve

# Problem 1: Multiples of 3 and 5

[Link to problem 1 ](https://projecteuler.net/problem=1)


In [2]:
s = 0
for i in range(1000):
    if i%3==0 or i%5==0:
        s+=i
s

233168

# Problem 2: Even Fibonacci numbers

[Link to problem 2 ](https://projecteuler.net/problem=2)

In [3]:
def fibo_stupid(n):
    if n==1:
        return 1
    if n==2:
        return 2
    return fibo_stupid(n-1) + fibo_stupid(n-2)
fibo = {1:1, 2:2}
def fibo_optimized(n):
    if n not in fibo:
        fibo[n] = fibo_optimized(n-1) + fibo_optimized(n-2)
    return fibo[n]

In [4]:
%%time
## we use the stupid method, notice also there is a beginner mistake which is calculating fibo_stupid two times instead of only one time
result = 0
n = 1
while fibo_stupid(n) < 4_000_000:
    n+=1
    if fibo_stupid(n)%2 == 0:
        result += fibo_stupid(n)
result

CPU times: user 3.54 s, sys: 20.2 ms, total: 3.56 s
Wall time: 3.59 s


4613732

In [5]:
%%time
## Calculating only one time the fibonnaci term
result = 0
n = 1
while True:
    fibo_term = fibo_stupid(n)
    if fibo_term>4_000_000:
        break
    if fibo_term%2 == 0:
        result += fibo_term
    n+=1
result 

CPU times: user 1.56 s, sys: 1.38 ms, total: 1.56 s
Wall time: 1.57 s


4613732

In [6]:
%%time
## Using the optimized fibonacci
while True:
    fibo_term = fibo_optimized(n)
    if fibo_term>4_000_000:
        break
    if fibo_term%2 == 0:
        result += fibo_term
    n+=1
result 

CPU times: user 19 µs, sys: 4 µs, total: 23 µs
Wall time: 24.1 µs


4613732

#### 25µs against 3.39s !!!!

# Problem 3: Largest prime factor

[Link to problem 3 ](https://projecteuler.net/problem=3)

Method 1
    
    Finding the divisors of n
    Filtering the primes divisors
    Choose the greatest prime divisor

Method 2
    
    Verify if n is prime, if not:
    In the search space (integers less than n/2) starting by the greatest check if that prime is a divisor of n and iterate back, break once one of them is found

In [7]:
def largest_prime_divisor(n):
    for divisor in sympy.divisors(n)[::-1]:
        if sympy.isprime(divisor):
            return divisor
largest_prime_divisor(600851475143)

6857

# Problem 4: Largest palindrome product

[Link to problem 4 ](https://projecteuler.net/problem=4)

In [8]:
def isPalindrome(n):
    if str(n) == str(n)[::-1]:
        return True
    return False
result = 0
for i in range(999,99,-1):
    for j in range(i,99,-1):
        if isPalindrome(i*j):
            if i*j>result:
                result = i*j
result

906609

# Problem 5: Largest palindrome product

[Link to problem 5 ](https://projecteuler.net/problem=5)

In [9]:
# Wet get the primes less than 20
def lcm(a, b):
    return abs(a*b) // math.gcd(a, b)
## Here comes the trick with reduce
L = list(range(1,21))
from functools import reduce
reduce(lcm, L)

232792560

# Problem 6: Sum square difference

[Link to problem 6 ](https://projecteuler.net/problem=6)

Method1: Mathematics we know what the sum of squares is and the sum of the arithmetic series.

Method2: iterate

In [10]:
def method1(n):
    return int((n*(n+1)/2)**2 - (n*(n+1)*(2*n+1))/6)
def method2(n):
    S1 = 0
    S2 = 0
    for i in range(n+1):
        S1 += i
        S2 += i**2
    return S1**2 - S2
method1(100), method2(100)

(25164150, 25164150)

# Problem 7: 10001st prime

[Link to problem 7 ](https://projecteuler.net/problem=7)

In [11]:
sieve.extend_to_no(10_001)
print(list(sieve._list)[10_000])

104743


# Problem 8: Largest product in a series

[Link to problem 8 ](https://projecteuler.net/problem=8)

In [12]:
nbr_str = """73167176531330624919225119674426574742355349194934
96983520312774506326239578318016984801869478851843
85861560789112949495459501737958331952853208805511
12540698747158523863050715693290963295227443043557
66896648950445244523161731856403098711121722383113
62229893423380308135336276614282806444486645238749
30358907296290491560440772390713810515859307960866
70172427121883998797908792274921901699720888093776
65727333001053367881220235421809751254540594752243
52584907711670556013604839586446706324415722155397
53697817977846174064955149290862569321978468622482
83972241375657056057490261407972968652414535100474
82166370484403199890008895243450658541227588666881
16427171479924442928230863465674813919123162824586
17866458359124566529476545682848912883142607690042
24219022671055626321111109370544217506941658960408
07198403850962455444362981230987879927244284909188
84580156166097919133875499200524063689912560717606
05886116467109405077541002256983155200055935729725
71636269561882670428252483600823257530420752963450"""
nbr_str = nbr_str.replace('\n', '')
list_numbers = [int(i) for i in list(nbr_str)]
result = 0
for i in range(len(list_numbers)-12):
    if np.product(list_numbers[i: i+13])>result:
        result = np.product(list_numbers[i: i+13])
result

23514624000

# Problem 9: Special Pythagorean triplet

[Link to problem 9 ](https://projecteuler.net/problem=9)

In [13]:
for a in range(1,1000):
    for b in range(a,1000):
        if 1000*(a+b)== a*b + 1000**2/2:
            print(a*b*(1000-a-b))
            break
    else:
        continue
    break

31875000


# Problem 10: Summation of primes

[Link to problem 10 ](https://projecteuler.net/problem=10)

In [14]:
primes = sieve.primerange(1, 2_000_000)
sum(primes)

142913828922

# Problem 11: Largest product in a grid

[Link to problem 11 ](https://projecteuler.net/problem=11)


In [15]:
list_str = """08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08
49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00
81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65
52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91
22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80
24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50
32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70
67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21
24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72
21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95
78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92
16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57
86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58
19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40
04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66
88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69
04 42 16 73 38 25 39 11 24 94 72 18 08 46 29 32 40 62 76 36
20 69 36 41 72 30 23 88 34 62 99 69 82 67 59 85 74 04 36 16
20 73 35 29 78 31 90 01 74 31 49 71 48 86 81 16 23 57 05 54
01 70 54 71 83 51 54 69 16 92 33 48 61 43 52 01 89 19 67 48"""
list_str = list_str.replace('\n', ' ')
L = [int(i) for i in list_str.split(' ')]
## AJOUTE CONGRUENCE
result = 0
for i in range(len(L)):
    #Horizontal
    if i%20<17:
        if L[i]*L[i+1]*L[i+2]*L[i+3]>result:
            result = L[i]*L[i+1]*L[i+2]*L[i+3]
    #Vertical
    if i<340:
        if L[i]*L[i+20]*L[i+40]*L[i+60]>result:
            result = L[i]*L[i+20]*L[i+40]*L[i+60]
    #Diagonal 1
    if i%20<17 and i<len(L)-63-1:
        if L[i]*L[i+21]*L[i+42]*L[i+63]>result:
            result = L[i]*L[i+21]*L[i+42]*L[i+63]
    #Diagonal 2
    if i%20>3 and i<len(L)-57-1:
        if L[i]*L[i+19]*L[i+38]*L[i+57]>result:
            result=L[i]*L[i+19]*L[i+38]*L[i+57]
result

70600674

# Problem 12: Highly divisible triangular number

[Link to problem 12 ](https://projecteuler.net/problem=12)

In [16]:
%%time
S=1
i=1
while True:
    if len(sympy.divisors(S))>=500:
        print(S)
        break
    i+=1
    S += i 

76576500
CPU times: user 552 ms, sys: 1.04 ms, total: 553 ms
Wall time: 553 ms


# Problem 13: Large sum

[Link to problem 13 ](https://projecteuler.net/problem=13)

In [17]:
nbr_str = """37107287533902102798797998220837590246510135740250
46376937677490009712648124896970078050417018260538
74324986199524741059474233309513058123726617309629
91942213363574161572522430563301811072406154908250
23067588207539346171171980310421047513778063246676
89261670696623633820136378418383684178734361726757
28112879812849979408065481931592621691275889832738
44274228917432520321923589422876796487670272189318
47451445736001306439091167216856844588711603153276
70386486105843025439939619828917593665686757934951
62176457141856560629502157223196586755079324193331
64906352462741904929101432445813822663347944758178
92575867718337217661963751590579239728245598838407
58203565325359399008402633568948830189458628227828
80181199384826282014278194139940567587151170094390
35398664372827112653829987240784473053190104293586
86515506006295864861532075273371959191420517255829
71693888707715466499115593487603532921714970056938
54370070576826684624621495650076471787294438377604
53282654108756828443191190634694037855217779295145
36123272525000296071075082563815656710885258350721
45876576172410976447339110607218265236877223636045
17423706905851860660448207621209813287860733969412
81142660418086830619328460811191061556940512689692
51934325451728388641918047049293215058642563049483
62467221648435076201727918039944693004732956340691
15732444386908125794514089057706229429197107928209
55037687525678773091862540744969844508330393682126
18336384825330154686196124348767681297534375946515
80386287592878490201521685554828717201219257766954
78182833757993103614740356856449095527097864797581
16726320100436897842553539920931837441497806860984
48403098129077791799088218795327364475675590848030
87086987551392711854517078544161852424320693150332
59959406895756536782107074926966537676326235447210
69793950679652694742597709739166693763042633987085
41052684708299085211399427365734116182760315001271
65378607361501080857009149939512557028198746004375
35829035317434717326932123578154982629742552737307
94953759765105305946966067683156574377167401875275
88902802571733229619176668713819931811048770190271
25267680276078003013678680992525463401061632866526
36270218540497705585629946580636237993140746255962
24074486908231174977792365466257246923322810917141
91430288197103288597806669760892938638285025333403
34413065578016127815921815005561868836468420090470
23053081172816430487623791969842487255036638784583
11487696932154902810424020138335124462181441773470
63783299490636259666498587618221225225512486764533
67720186971698544312419572409913959008952310058822
95548255300263520781532296796249481641953868218774
76085327132285723110424803456124867697064507995236
37774242535411291684276865538926205024910326572967
23701913275725675285653248258265463092207058596522
29798860272258331913126375147341994889534765745501
18495701454879288984856827726077713721403798879715
38298203783031473527721580348144513491373226651381
34829543829199918180278916522431027392251122869539
40957953066405232632538044100059654939159879593635
29746152185502371307642255121183693803580388584903
41698116222072977186158236678424689157993532961922
62467957194401269043877107275048102390895523597457
23189706772547915061505504953922979530901129967519
86188088225875314529584099251203829009407770775672
11306739708304724483816533873502340845647058077308
82959174767140363198008187129011875491310547126581
97623331044818386269515456334926366572897563400500
42846280183517070527831839425882145521227251250327
55121603546981200581762165212827652751691296897789
32238195734329339946437501907836945765883352399886
75506164965184775180738168837861091527357929701337
62177842752192623401942399639168044983993173312731
32924185707147349566916674687634660915035914677504
99518671430235219628894890102423325116913619626622
73267460800591547471830798392868535206946944540724
76841822524674417161514036427982273348055556214818
97142617910342598647204516893989422179826088076852
87783646182799346313767754307809363333018982642090
10848802521674670883215120185883543223812876952786
71329612474782464538636993009049310363619763878039
62184073572399794223406235393808339651327408011116
66627891981488087797941876876144230030984490851411
60661826293682836764744779239180335110989069790714
85786944089552990653640447425576083659976645795096
66024396409905389607120198219976047599490197230297
64913982680032973156037120041377903785566085089252
16730939319872750275468906903707539413042652315011
94809377245048795150954100921645863754710598436791
78639167021187492431995700641917969777599028300699
15368713711936614952811305876380278410754449733078
40789923115535562561142322423255033685442488917353
44889911501440648020369068063960672322193204149535
41503128880339536053299340368006977710650566631954
81234880673210146739058568557934581403627822703280
82616570773948327592232845941706525094512325230608
22918802058777319719839450180888072429661980811197
77158542502016545090413245809786882778948721859617
72107838435069186155435662884062257473692284509516
20849603980134001723930671666823555245252804609722
53503534226472524250874054075591789781264330331690"""
list_numbers = [int(i) for i in nbr_str.split('\n')]
sum(list_numbers)//10**(len(str(sum(list_numbers)))-10)

5537376230

# Problem 14: Longest Collatz sequence

[Link to problem 14 ](https://projecteuler.net/problem=14)

In [18]:
def collatz_sequence(n):
    L = [n]
    while n != 1:
        if n%2 == 0:
            L.append(n//2)
            n = n//2
        else:
            L.append(3*n+1)
            n = 3*n + 1
    return L
L = dict()
for i in range(1, 1_000_000):
    L[i] = len(collatz_sequence(i))
max(L, key=L.get)

837799

# Problem 15: Lattice paths

[Link to problem 15 ](https://projecteuler.net/problem=15)


The number of different permutations of $n$ objects, where there are $n_1$ indistinguishable objects of type 1,$n_2$ indistinguishable objects of type 2,. . ., and $n_k$ indistinguishable objects of type $k$, is:
     

$$ \frac{n!}{n_{1}!n_{2}!....n_{k}!} $$


In [19]:
from math import factorial
print(factorial(20*2)//(factorial(20)*factorial(20)))

137846528820


# Problem 16: Power digit sum

[Link to problem 16 ](https://projecteuler.net/problem=16)

In [20]:
nbr_list = [int(i) for i in str(2**1000)]
print(sum(list(nbr_list)))

1366


# Problem 17: Number letter counts

[Link to problem 17 ](https://projecteuler.net/problem=17)

In [21]:
from num2words import num2words
import re
Count = 0
for i in range(1,1_001):
    Count += sum([len(element) for element in re.findall(r"[\w']+", num2words(i))])
Count

21124

# Problem 18: Maximum path sum I

[Link to problem 18 ](https://projecteuler.net/problem=18)

In [22]:
triangle_str='''75
95 64
17 47 82
18 35 87 10
20 04 82 47 65
19 01 23 75 03 34
88 02 77 73 07 63 67
99 65 04 28 06 16 70 92
41 41 26 56 83 40 80 70 33
41 48 72 33 47 32 37 16 94 29
53 71 44 65 25 43 91 52 97 51 14
70 11 33 28 77 73 17 78 39 68 17 57
91 71 52 38 17 14 91 43 58 50 27 29 48
63 66 04 68 89 53 67 30 73 16 69 87 40 31
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23'''
triangle = [[int(j) for j in i.split()] for i in triangle_str.split("\n")]
triangle.reverse()

def update_line(i):
    for k in range(len(triangle[i])):
        triangle[i][k] = triangle[i][k] + max(triangle[i-1][k], triangle[i-1][k+1])

for i in range(1,15):
    update_line(i)
triangle[-1][0]

1074

# Problem 19: Counting Sundays

[Link to problem 19 ](https://projecteuler.net/problem=19)

In [23]:
#Solution 1
list_days_leap = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
list_days_not = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
Days_new = {6: "Monday", 0: "Tuesday", 1: "Wednesday", 2 : "Thursday", 3: "Friday", 4: "Saturday", 5: "Sunday"}
sum_days = 0
nbr_sundays = 0
for year in range(1901, 2001):
    if year%4 == 0:
        list_days = list_days_leap
    else:
        list_days = list_days_not
    for month in list_days:
        sum_days += month
        if sum_days% 7== 5:
            nbr_sundays += 1
nbr_sundays

171

In [24]:
#Solution 2
import datetime
nbr_sundays2 = 0
for year in range(1901, 2001):
    for month in range(1,13):
        if datetime.datetime(year, month, 1).weekday() == 6:
            nbr_sundays2 += 1
nbr_sundays2

171

# Problem 20: Factorial digit sum

[Link to problem 20 ](https://projecteuler.net/problem=20)

In [25]:
def sum_factorial(n):
    return sum([int(i) for i in list(str(factorial(n)))])
sum_factorial(100)

648

# Problem 21: Amicable numbers

[Link to problem 21 ](https://projecteuler.net/problem=21)

In [26]:
# Solution 1 
def list_divisors(n):
    divisors = []
    for i in range(1, int(np.sqrt(n))+1):
        if n%i == 0:
            if n//i >= i:
                divisors.append(i)
                if n//i > i:
                    divisors.append(n//i)
    return divisors
def sum_proper_divisors(n):
    list_div = list_divisors(n)
    list_div.remove(n)
    return sum(list_div)
Somme = 0
for a in range(2,10001):
    b = sum_proper_divisors(a)
    if sum_proper_divisors(b) == a and a!=b:
        Somme += a
Somme

31626

In [27]:
# Solution 2
Somme = 0
for a in range(2,10001):
    b = sum(sympy.divisors(a)[:-1])
    if sum(sympy.divisors(b)[:-1]) == a and a!=b:
        Somme += a
Somme

31626

# Problem 22: Names scores

[Link to problem 22 ](https://projecteuler.net/problem=22)

In [28]:
import pandas as pd
fichier = open("names.txt", "r")
names = list(eval(fichier.read()))
fichier.close()
data = pd.DataFrame(data={'names':names})
data = data.sort_values('names')
data.reset_index(inplace=True, drop=True)
def word_value(word):
    return sum([ord(char) - 96 for char in word.lower()])
data['value_name'] = data.names.apply(lambda x: word_value(x))
def sum_values(index, value_name):
    products = [a * b for a, b in zip(index, value_name)]
    return sum(products)
sum_values(list(data.index+1), list(data.value_name))

871198282

# Problem 23: Non-abundant sums

[Link to problem 23 ](https://projecteuler.net/problem=23)

In [29]:
from collections import Counter
def list_divisors(n):
    divisors = []
    for i in range(1, int(np.sqrt(n))+1):
        if n%i == 0:
            if n//i >= i:
                divisors.append(i)
                if n//i > i:
                    divisors.append(n//i)
    return divisors
def sum_proper_divisors(n):
    list_div = list_divisors(n)
    list_div.remove(n)
    return sum(list_div)
def find_abundant_numbers(n):
    out = []
    for i in range(1,n):
        if sum_proper_divisors(i) > i:
            out += [i]
    return out
abundant_numbers = find_abundant_numbers(28124)
possible_sums = []
for i in range(len(abundant_numbers)):
    for j in range(i, len(abundant_numbers)):
        somme = abundant_numbers[i] + abundant_numbers[j]
        if somme > 28123:
            break
        possible_sums += [somme]
resultat = (28123*28124)//2 - sum(Counter(possible_sums).keys())
resultat

4179871

# Problem 24: Lexicographic permutations

[Link to problem 24 ](https://projecteuler.net/problem=24)

In [30]:
## First solution by finding the possible position of each digit
999998 == 2*factorial(9) + 6*factorial(8) + 6*factorial(7) + 2*factorial(6) + 5*factorial(5) + 1*factorial(4) + 2*factorial(3) + factorial(2)
result1 = 2783915460

## second solution
import itertools
permut = list(itertools.permutations("0123456789"))[999999]
result2 = ''.join(permut)
result1, int(result2)

(2783915460, 2783915460)

# Problem 25: 1000-digit Fibonacci number

[Link to problem 25 ](https://projecteuler.net/problem=25)

In [31]:
a = 1
b = 1
resultat = 2
while len(str(b)) < 1000:
    a, b = b, a+b
    resultat += 1
print(resultat)

4782


# Problem 26: Reciprocal cycles

[Link to problem 26 ](https://projecteuler.net/problem=26)

In [32]:
def cycle_length(den):
    reste = 10
    i = 0
    #Calcul des décimales tant que le reste n'est pas égal à 10
    while reste != 10 or i < 1:
        reste = (reste % den) * 10
        i += 1
    return i
longest = 0
for i in range(2, 1000):
    if i%2 != 0 and i%5 != 0:
        length = cycle_length(i)
        if length > longest:
            longest = length
            resultat = i
print(resultat)

983


# Problem 27: Quadratic primes

[Link to problem 27 ](https://projecteuler.net/problem=27)

In [33]:
from sympy import sieve
import sympy
sieve._reset()
sieve.extend_to_no(170)
a = list(sieve._list)
a = [i for i in a if i<1000]
a += [-i for i in a]
def calculate(a,b):
    n=0
    while sympy.isprime(n**2+n*a+b):
        n+=1
    return n
max_n = 0
for i in range(-499, 500):
    for b in a:
        if calculate(2*i+1, b)>max_n:
            max_n = calculate(2*i+1, b)
            a_out = 2*i+1
            b_out = b
a_out*b_out     

-59231

# Problem 28: Number spiral diagonals

[Link to problem 28 ](https://projecteuler.net/problem=28)

In [34]:
n=1
somme = 1
counter = 0
i=0
while i<500:
    while counter < 4:
        n = n + 2*(i+1)
        somme+= n
        
        counter += 1
    i+=1
    counter = 0
somme

669171001

# Problem 29 Distinct powers

[Link to problem 29](https://projecteuler.net/problem=29)

In [35]:
# Problem 29 Distinct powers
distinct_powers = dict()
for a in range(2,101):
    for b in range(2,101):
        distinct_powers[a**b] = 1
len(distinct_powers)

9183

# Problem 30 Digit fifth powers

[Link to problem 30](https://projecteuler.net/problem=30)


Let $ m = \overline{a_{n}a_{n-1}.......a_{0}} = \sum_{i=0}^{i=n}a_{i}^{5}$

Knowing that $0 \leq a_{i}\leq 9$ we have $\sum_{i=0}^{i=n}a_{i}^{5} <= 59049n$

Moreover we have $a_{n}>=1$ and that $10^{6}-59049*6 > 0$

The search domain is for $n \leq 4$ that means in $m<10^{6}$

In [36]:
out = []
for i in range(2,10**6):
    if i == sum([int(j)**5 for j in list(str(i))]):
        out += [i]
sum(out)

443839

# Problem 31 Coin sums

[Link to problem 31](https://projecteuler.net/problem=31)

In [37]:
resultat = 0
reste = 200
for a in range(0, reste+1, 200):
    reste = 200 - a
    for b in range(0, reste+1, 100):
        reste = 200 - a - b
        for c in range(0, reste+1, 50):
            reste = 200 - a - b - c
            for d in range(0, reste+1, 20):
                reste = 200 - a - b - c - d
                for e in range(0, reste+1, 10):
                    reste = 200 - a - b - c - d - e
                    for f in range(0, reste+1, 5):
                        reste = 200 - a - b - c - d - e - f
                        for g in range(0, reste+1, 2):
                            h = 200 - a - b - c - d - e - f - g
                            if h >= 0:
                                resultat += 1
resultat

73682

# Problem 32 Pandigital products

[Link to problem 32](https://projecteuler.net/problem=32)

In [38]:
#### Problem 32
from itertools import permutations
L = list(permutations('123456789'))
L = [''.join(i) for i in L]
S = set()
for k in range(len(L)):
    for i in range(1,9):
        for j in range(i+1,9):
            x = int(L[k][:i])
            y = int(L[k][i:j])
            z = int(L[k][j:])
            if x<y<z:
                if x*y == z:
#                     print(x,y,z)
                    S.add(z)
print(sum(S))

45228


# Problem 33 Digit cancelling fractions

[Link to problem 33](https://projecteuler.net/problem=33)

In [39]:
L = []
for a in range(1,10):
    for b in range(1,10):
        for c in range(1,10):
            for d in range(1,10):
                actual = int(str(a)+str(b)) / int(str(c)+str(d))
                if actual <1:
                    if a/c == actual and b==d:
                        L.append([str(a)+str(b), str(c)+str(d)])
                    if a/d == actual and b==c:
                        L.append([str(a)+str(b), str(c)+str(d)])
                    if b/d == actual and a==c:
                        L.append([str(a)+str(b), str(c)+str(d)])
                    if b/c == actual and a==d:
                        L.append([str(a)+str(b), str(c)+str(d)])
L
# the product of these four fractions is given in its lowest common terms gives
result = 5*4*5
result

100

# Problem 34 Digit factorials

[Link to problem 34](https://projecteuler.net/problem=34)


Let $ m = \overline{a_{n}a_{n-1}.......a_{0}} = \sum_{i=0}^{i=n}a_{i}!$

Knowing that $0 \leq a_{i}\leq 9$ Moreover we have $a_{n}>=1$ 

$\sum_{i=0}^{i=n}a_{i}! \leq 9!n$ and $m \geq 10^{n}$

and we can check that $10^6-9!x6 < 0$ and $10^7-9!x7 > 0$

In [40]:
L = []
for k in range(3,10**7):
    if sum([factorial(int(i)) for i in list(str(k))]) == k:
        L +=[k]
sum(L)

40730

# Problem 35 Circular primes

[Link to problem 35](https://projecteuler.net/problem=35)

In [41]:
## A way to creates all possible permutations for a number
def permuation_number(n):
    l = list(str(n))
    return list(set([int(''.join(i)) for i in [[l[i - j] for i in range(len(l))] for j in range(len(l))]]))

from collections import defaultdict
L = defaultdict(int)
for i in range(2,10**6):
    if i==2 or i==5:
        L[i] = 1
    d = i%10
    if (d%2 == 0) or d==5:
        continue
    else:
        if L[i] == 0:
            permu = permuation_number(i)
            j = 0
            for k in permu:
                if sympy.isprime(k):
                    j += 1
                else:
                    break
            if j == len(permu):
                for k in permu:
                    L[k] = 1
sum(L.values())

55

# Problem 36 Double-base palindromes

[Link to problem 36](https://projecteuler.net/problem=36)

In [42]:
def isPalindrome(n):
    return str(n) == str(n)[::-1]
L=[]
for i in range(10**6):
    if isPalindrome(i) and isPalindrome(int(bin(i)[2:])):
        L += [i]
sum(L)

872187

# Problem 37 Truncatable primes

[Link to problem 37](https://projecteuler.net/problem=37)

In [43]:
def prime_verification(n):
    if n==1:
        return False
    if n==2:
        return True
    if n==5:
        return True
    if ((n%10)%2 == 0) or ((n%10) == 5):
        return False

    for i in range(2,ceil(np.sqrt(n))+1):
        if n%i == 0:
            return False
    return True

def trunctated(n):
    a = str(n)
    lon = len(str(n))
    return [int(i) for i in [a[i:] for i in range(len(a))] + [a[:i] for i in range(1,len(a))]]
i = 0
n = 10
L = []
while i<11:
    if prime_verification(n):
        j=0
        l = trunctated(n)
        for k in l:
            if prime_verification(k):
                j+=1
        if j == len(l):
            L+=[n]
            i+=1
    n+=1
sum(L)

748317

# Problem 38 Pandigital multiples

[Link to problem 38](https://projecteuler.net/problem=38)

Starting by limiting the search space. 
    
        The number of digits cannot exceed 5 digits since n>1
        A given information is that one candidate starts with 9, that gives us a hint that the number we are looking for starts with 9.
        If the number of digits is 2, for 1 we have 2, for 2 we have 3 for 3 we have 3 for 4 we have 3: that means for n=3 8 digits number and n=4 11 digit number (we are looking for 9 digit number)
        If the number of digits is 3, for 1 we have 3, for 2 we have 4, and for 3 we have 4 that means n=3 gives 11 digits number (again false)
That leaves us with 4 digit number with this form 9***.

We notice also that 98** , 97** , 96** and 95** multiplied by 2 gives a 19*** with a 9in the digits of the product but we already have 9.

That leaves us with 94** ,93** ,92**, 91**

We can add that for the case of 91**, if we do 91** x 1 = 91** and 91** x 2 = 1**** it means 1 repeated two times

That leaves us with 94** , 93** , 92** , that's 123 possible solutions


In [44]:
s_3 = ['1','2','4','5','6','7','8']
s_2 = ['1','3','4','5','6','7','8']
s_4 = ['1','2','3','5','6','7','8']
from itertools import permutations
a_4 = '94'
a_3 = '93'
a_2 = '92'
def verify(a_4, s_4):
    L = [''.join(i) for i in list(permutations(s_4,2))]
    for element in L:
        out = a_4 + element + str(int(a_4 + element)*2)
        if len(list(out)) == len(set(out)):
            print(out, a_4+element)
verify(a_4,s_4)
verify(a_3,s_3)
verify(a_2,s_2)

932718654 9327
935218704 9352
923518470 9235
926718534 9267
927318546 9273


The answer is obviously 932718654

# Problem 39 Integer right triangles

[Link to problem 39](https://projecteuler.net/problem=39)

a+b+c = p  and a²+b²=c², change c by p-a-b

In [ ]:
def find_solution(p):
    L=[]
    for a in range(1,p):
        for b in range(a,p):
            if 2*p*(a+b) == p**2+2*a*b:
                L += [(a,b)]
    return L
m = 0
pmax = 2
for p in range(2,1001):
    if len(find_solution(p)) > m:
        m = len(find_solution(p))
        pmax=p
pmax

# Problem 40 Champernowne's constant

[Link to problem 40](https://projecteuler.net/problem=40)

In [ ]:
##### Problem 40
L = [1, 10, 100, 1_000, 10_000, 100_000, 1_000_000]
out = []
i=0
j=1
while i < max(L):
    n=len(str(j))
    for k in range(1,n+1):
        if i+k in L:
            out += [int(str(j)[k-1])]
    j+=1
    i+=n
np.product(out)    

# Problem 41 Pandigital prime

[Link to problem 41](https://projecteuler.net/problem=41)

In [ ]:
def prime_verification(n):
    if n==2:
        return True
    
    if ((n%10)%2 == 0) or ((n%10) == 5):
        return False
    if n==5:
        return True
    for i in range(2,ceil(np.sqrt(n))+1):
        if n%i == 0:
            return False
    return True
for j in range(10,1,-1):
    L = list(permutations(list(range(1,j))))
    L = [int(''.join(map(str, element))) for element in L]
    L.sort(reverse=True)
    for i in range(len(L)):
        if prime_verification(L[i]):
            print(L[i])
            break
    else:
        continue
    break

# Problem 42 Coded triangle numbers

[Link to problem 42](https://projecteuler.net/problem=42)

In [ ]:
import pandas as pd
import sys
fichier = open("words.txt", "r")
words = list(eval(fichier.read()))
fichier.close()
data = pd.DataFrame(data={'words':words})
# Trick without solving the second degreen equation and see if there are integers solution, we inspect the data
len_max = data.words.apply(len).max()
value_max = len_max*26
def evaluate(word):
    return sum([ord(char) - 96 for char in word.lower()])
data['value'] = data.words.apply(evaluate)
def t_n(n):
    return int((n*(n+1))/2)
t = []
j=1
while t_n(j)<365:
    t+=[t_n(j)]
    j+=1
def verification_triangle(row):
    if row.value in t:
        return 1
    else:
        return 0
data['verif']= data.apply(verification_triangle, axis=1)
data['verif'].sum()

# Problem 43 Sub-string divisibility

[Link to problem 43](https://projecteuler.net/problem=43)

In [ ]:
L = list(permutations(list(range(10))))
L = [''.join(map(str, element)) for element in L]
def joining(n,i,j,k):
    return int(n[i-1]+n[j-1]+n[k-1])
def verfication_property(n):
    if (joining(n,2,3,4)%2==0) and (joining(n,3,4,5)%3==0) and (joining(n,4,5,6)%5==0) and (joining(n,5,6,7)%7==0)and (joining(n,5,6,7)%7==0) and (joining(n,6,7,8)%11==0) and (joining(n,7,8,9)%13==0) and (joining(n,8,9,10)%17==0):
        return True
    return False
s=0
for element in L:
    if verfication_property(element):
        s+=int(element)
s

# Problem 44 Pentagon numbers

[Link to problem 44](https://projecteuler.net/problem=44)

In [ ]:
def p_n(n):
    return int(n*(3*n-1))/2
for i in range(1,3000):
    for j in range(i,3001):
        a = (np.sqrt(24*(p_n(j) + p_n(i))+1)+1)/6
        b = (np.sqrt(24*(p_n(j) - p_n(i))+1)+1)/6
        if a == int(a) and b==int(b):
            print(int(p_n(j)- p_n(i)))
            break
    else:
        continue
    break

# Problem 45 Triangular, pentagonal, and hexagonal

[Link to problem 45](https://projecteuler.net/problem=45)

In [ ]:
def h_n(n):
    return n*(2*n-1)
#h_n(143) == 40755
i = 144
a= (np.sqrt(8*h_n(i)+1)-1)/2
b= (np.sqrt(24*h_n(i)+1)+1)/6
while (int(a)!= a) or (int(b)!=b):
    i+=1
    a= (np.sqrt(8*h_n(i)+1)-1)/2
    b= (np.sqrt(24*h_n(i)+1)+1)/6
print(h_n(i))

# Problem 46 Goldbach's other conjecture

[Link to problem 46](https://projecteuler.net/problem=46)

In [ ]:
def prime_verification(n):
    if n==2:
        return True
    if n==5:
        return True
    
    if ((n%10)%2 == 0) or ((n%10) == 5):
        return False

    for i in range(2,ceil(np.sqrt(n))+1):
        if n%i == 0:
            return False
    return True

def prime_search(n):
    l = []
    for i in range(n):
        if prime_verification(i):
            l += [i]
    return l
def goldbach_verification(n):
    l = prime_search(n-1)
    for prime in l:
        if (n-prime)/2 == (n-prime)//2 and np.sqrt((n-prime)/2) == int(np.sqrt((n-prime)/2)):
            return True
    return False
i=33
while goldbach_verification(i):
    i+=2
    while prime_verification(i):
        i+=2
i

# Problem 47 Distinct primes factors

[Link to problem 47](https://projecteuler.net/problem=47)

In [ ]:
def npf(number):
    """function which will return
    the number of prime factors"""
    i = 2
    a = set()
    while i < number**0.5 or number == 1:
        if number % i == 0:
            number = number/i
            a.add(i)
            i -= 1
        i += 1
    return (len(a)+1)
i=3*5*7*2
while True:
    if npf(i)==4:
        if npf(i+1)==4:
            if npf(i+2)==4:
                if npf(i+3)==4:
                    print(i)
                    break
                else:
                    i+=1
            else:
                i+=1
        else:
            i+=1
    else:
        i+=1

# Problem 48 Self powers

[Link to problem 48](https://projecteuler.net/problem=48)

In [ ]:
def sum_calcu(n):
    s=0
    for i in range(1,n+1):
        s += i**i
    return s
sum_calcu(1000)%(10**10)

# Problem 49 Prime permutations

[Link to problem 49](https://projecteuler.net/problem=49)

In [ ]:
from itertools import permutations
def check_premutation(n,m):
    if m in [int(''.join(element)) for element in list(permutations(str(n)))]:
        return True
    return False
def prime_verification(n):
    if n==2:
        return True
    if n==5:
        return True
    
    if ((n%10)%2 == 0) or ((n%10) == 5):
        return False

    for i in range(2,ceil(np.sqrt(n))+1):
        if n%i == 0:
            return False
    return True
l=[]
for i in range(1_000,10_000-3330*2):
    if prime_verification(i):
        if prime_verification(i+3330):
            if check_premutation(i, i+3330):
                if prime_verification(i+3330*2):
                    if check_premutation(i,i+3330*2):
                        l+=[i]
int(str(l[1])+str(l[1]+3330)+str(l[1]+3330*2))

# Problem 50 Consecutive prime sum

[Link to problem 50](https://projecteuler.net/problem=50)

In [ ]:
from sympy import sieve
import sympy
sieve._reset()
sieve.extend_to_no(4000)
a = list(sieve._list)[:-784]

L = dict()
out = 0
n= 0 
for i in range(len(a)-1):
    for k in range(i+1,len(a)):
        if sum(a[i:k])>1_000_000:
            break
            
        if sympy.isprime(sum(a[i:k])):
            L[i] = k
            if len(a[i:k])>n:
                n= len(a[i:k])
                out = sum(a[i:k])
        

n, out